// @hemp-trunfo/engine/dist/cartas/baralho.js
export const baralhoCompleto = [
    {
        id: 'A1', nome: 'Clockwork Orange', banco: 'Crazy Seeds',
        descricao: 'Homenagem ao clássico time holandês de 1972. Terpenos cítricos, cheiro forte de laranja.',
        thc: 21, cbd: 0.03, relaxamento: 25, foco: 91, felicidade: 41, fome: 59, sono: 20,
        tipo: 'genetica'
    },
    {
        id: 'A2', nome: 'Monkey Fever', banco: 'Crazy Seeds',
        descricao: 'A "Febre do Macaco" tem sabor de uva passa. Genética 100% Indica, Afghan com Gorilla.',
        thc: 21, cbd: 0.04, relaxamento: 94, foco: 47, felicidade: 63, fome: 81, sono: 100,
        tipo: 'genetica'
    },
    {
        id: 'A3', nome: 'Stinky Fat', banco: 'Crazy Seeds',
        descricao: 'Cheiro de queijo azedo. Flores densas, ótima para amarração.',
        thc: 25, cbd: 0.11, relaxamento: 57, foco: 69, felicidade: 90, fome: 77, sono: 55,
        tipo: 'genetica'
    },
    {
        id: 'A4', nome: 'Crazy Haze', banco: 'Crazy Seeds',
        descricao: 'Cepa pegajosa. Nome significa "Névoa Maluca", flores podem ficar roxas.',
        thc: 21, cbd: 0.01, relaxamento: 100, foco: 27, felicidade: 75, fome: 84, sono: 93,
        tipo: 'genetica'
    },
    {
        id: 'A5', nome: 'Black Harpia', banco: 'Crazy Seeds',
        descricao: 'Genética brasileira com flores pretas. Nome faz jus ao "Gavião Real".',
        thc: 26, cbd: 0.03, relaxamento: 70, foco: 63, felicidade: 71, fome: 55, sono: 70,
        tipo: 'genetica'
    },
    {
        id: 'A6', nome: 'Purple Dream', banco: 'Crazy Seeds',
        descricao: 'Flores roxas intensas. Relaxante profundo com toques criativos.',
        thc: 24, cbd: 0.05, relaxamento: 88, foco: 45, felicidade: 82, fome: 70, sono: 85,
        tipo: 'genetica'
    },
    {
        id: 'A7', nome: 'Green Goblin', banco: 'Crazy Seeds',
        descricao: 'Energia explosiva e foco laser. Para tarefas criativas.',
        thc: 23, cbd: 0.02, relaxamento: 15, foco: 95, felicidade: 78, fome: 65, sono: 10,
        tipo: 'genetica'
    },
    {
        id: 'A8', nome: 'Golden Kush', banco: 'Crazy Seeds',
        descricao: 'Tricomas dourados. Sabor terroso com notas cítricas.',
        thc: 22, cbd: 0.08, relaxamento: 65, foco: 55, felicidade: 70, fome: 90, sono: 60,
        tipo: 'genetica'
    },
    {
        id: 'A9', nome: 'Silver Haze', banco: 'Crazy Seeds',
        descricao: 'Tricomas prateados. Efeito cerebral energético.',
        thc: 20, cbd: 0.04, relaxamento: 40, foco: 85, felicidade: 75, fome: 50, sono: 30,
        tipo: 'genetica'
    },
    {
        id: 'A10', nome: 'Red Dragon', banco: 'Crazy Seeds',
        descricao: 'Flores avermelhadas, aroma picante. Potência elevada.',
        thc: 28, cbd: 0.01, relaxamento: 80, foco: 35, felicidade: 85, fome: 95, sono: 75,
        tipo: 'genetica'
    },
    {
        id: 'A11', nome: 'Blueberry Kush', banco: 'Crazy Seeds',
        descricao: 'Sabor de mirtilo, relaxante corporal intenso.',
        thc: 19, cbd: 0.06, relaxamento: 92, foco: 30, felicidade: 88, fome: 85, sono: 90,
        tipo: 'genetica'
    },
    {
        id: 'A12', nome: 'White Widow', banco: 'Crazy Seeds',
        descricao: 'Tricomas brancos como neve. Equilíbrio corpo e mente.',
        thc: 20, cbd: 0.03, relaxamento: 60, foco: 72, felicidade: 80, fome: 60, sono: 45,
        tipo: 'genetica'
    },
    {
        id: 'A13', nome: 'Northern Lights', banco: 'Crazy Seeds',
        descricao: 'Lendas dizem que veio das estrelas. Sabor doce, sedativo.',
        thc: 18, cbd: 0.02, relaxamento: 95, foco: 20, felicidade: 70, fome: 80, sono: 95,
        tipo: 'genetica'
    },
    {
        id: 'A14', nome: 'Sour Diesel', banco: 'Crazy Seeds',
        descricao: 'Aroma de combustível cítrico. Efeito cerebral energético.',
        thc: 22, cbd: 0.01, relaxamento: 30, foco: 90, felicidade: 85, fome: 45, sono: 15,
        tipo: 'genetica'
    },
    {
        id: 'A15', nome: 'OG Kush', banco: 'Crazy Seeds',
        descricao: 'Clássico da Costa Oeste. Limão, pinho e especiarias.',
        thc: 24, cbd: 0.04, relaxamento: 75, foco: 50, felicidade: 72, fome: 88, sono: 65,
        tipo: 'genetica'
    },
    {
        id: 'A16', nome: 'Amnesia Haze', banco: 'Crazy Seeds',
        descricao: 'Premiada, sabor cítrico e floral. Efeito cerebral intenso.',
        thc: 23, cbd: 0.02, relaxamento: 35, foco: 88, felicidade: 90, fome: 55, sono: 25,
        tipo: 'genetica'
    },
    {
        id: 'A17', nome: 'Granddaddy Purple', banco: 'Crazy Seeds',
        descricao: 'Flores roxas profundas, aroma de uva. Relaxante poderoso.',
        thc: 17, cbd: 0.05, relaxamento: 98, foco: 15, felicidade: 65, fome: 92, sono: 100,
        tipo: 'genetica'
    },
    {
        id: 'A18', nome: 'Jack Herer', banco: 'Crazy Seeds',
        descricao: 'Homenagem ao ativista. Pinho cítrico, criatividade e clareza.',
        thc: 21, cbd: 0.03, relaxamento: 45, foco: 93, felicidade: 87, fome: 50, sono: 20,
        tipo: 'genetica'
    },
    {
        id: 'A19', nome: 'Girl Scout Cookies', banco: 'Crazy Seeds',
        descricao: 'Doce de menta e chocolate. Equilibrado com toque de euforia.',
        thc: 25, cbd: 0.02, relaxamento: 70, foco: 60, felicidade: 92, fome: 85, sono: 55,
        tipo: 'genetica'
    },
    {
        id: 'A20', nome: 'Gorilla Glue', banco: 'Crazy Seeds',
        descricao: 'Tricomas pegajosos. Potência extrema, efeito pesado.',
        thc: 27, cbd: 0.01, relaxamento: 85, foco: 25, felicidade: 75, fome: 90, sono: 80,
        tipo: 'genetica'
    },
    {
        id: 'A21', nome: 'Pineapple Express', banco: 'Crazy Seeds',
        descricao: 'Sabor tropical de abacaxi. Efeito energético e alegre.',
        thc: 20, cbd: 0.04, relaxamento: 50, foco: 80, felicidade: 95, fome: 70, sono: 30,
        tipo: 'genetica'
    },
    {
        id: 'A22', nome: 'AK-47', banco: 'Crazy Seeds',
        descricao: 'Híbrida equilibrada, sabor terroso e floral. Social e criativa.',
        thc: 19, cbd: 0.03, relaxamento: 55, foco: 75, felicidade: 80, fome: 65, sono: 40,
        tipo: 'genetica'
    },
    {
        id: 'A23', nome: 'Chemdawg', banco: 'Crazy Seeds',
        descricao: 'Aroma químico único. Efeito cerebral intenso e misterioso.',
        thc: 24, cbd: 0.01, relaxamento: 40, foco: 85, felicidade: 70, fome: 60, sono: 35,
        tipo: 'genetica'
    },
    {
        id: 'A24', nome: 'Mango Kush', banco: 'Crazy Seeds',
        descricao: 'Sabor doce de manga tropical. Relaxante e feliz.',
        thc: 18, cbd: 0.06, relaxamento: 78, foco: 40, felicidade: 88, fome: 95, sono: 70,
        tipo: 'genetica'
    },
    {
        id: 'A25', nome: 'Lemon Haze', banco: 'Crazy Seeds',
        descricao: 'Sabor cítrico fresco de limão. Cerebral claro e energético.',
        thc: 21, cbd: 0.03, relaxamento: 30, foco: 92, felicidade: 86, fome: 55, sono: 20,
        tipo: 'genetica'
    },
    {
        id: 'A26', nome: 'Bubble Gum', banco: 'Crazy Seeds',
        descricao: 'Sabor doce de chiclete rosa. Feliz e descontraído.',
        thc: 17, cbd: 0.04, relaxamento: 62, foco: 55, felicidade: 93, fome: 75, sono: 50,
        tipo: 'genetica'
    },
    {
        id: 'A27', nome: 'Strawberry Cough', banco: 'Crazy Seeds',
        descricao: 'Sabor doce de morango. Cerebral sociável.',
        thc: 19, cbd: 0.02, relaxamento: 42, foco: 78, felicidade: 89, fome: 58, sono: 28,
        tipo: 'genetica'
    },
    {
        id: 'A28', nome: 'Durban Poison', banco: 'Crazy Seeds',
        descricao: 'Landrace da África do Sul. Anis, energético puro.',
        thc: 22, cbd: 0.01, relaxamento: 20, foco: 96, felicidade: 84, fome: 40, sono: 12,
        tipo: 'genetica'
    },
    {
        id: 'VANTAGEM', nome: 'Vantagem', tipo: 'vantagem', efeito: 'vence_automatico'
    },
    {
        id: 'REVES', nome: 'Revés', tipo: 'reves', efeito: 'perde_automatico'
    },
    {
        id: 'INFO', nome: 'Breve História da Cannabis', tipo: 'informativa', efeito: 'informativa'
    }
];
export function embaralhar(cartas) {
    const copia = [...cartas];
    for (let i = copia.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [copia[i], copia[j]] = [copia[j], copia[i]];
    }
    return copia;
}
export function filtrarCartasJogaveis(cartas) {
    return cartas.filter(c => c.tipo === 'genetica' || c.tipo === 'vantagem' || c.tipo === 'reves');
}
